common/scala/src/main/scala/whisk/core/entity/InvokerInstanceId.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.entity
+
+import spray.json.DefaultJsonProtocol
+
+case class InvokerInstanceId(val instance: Int, name: Option[String] = None) {
+  def toInt: Int = instance
+}
+
+case class ControllerInstanceId(private val instance: String) {
```

Comment:
if we don't match on the case class anywhere, might as well make this as regular class which will actually hide  `instance`.

Reply:
I tend to disagree, case classes provide equals and everything.

Can we name `instance` to `asString`? Then you can make it public and leave it as is.

Reply:
Also instance is guarded because it should not be used as it hasn’t been sanitized. The intent is to only expose asString. An alternate is to move the require and cleaning to “replaced” into a singleton constructor method so that it’s not possible to retrieve the original string.

Reply:
Should it be a [value class](https://docs.scala-lang.org/overviews/core/value-classes.html) then? Would give us the benefit of 0 allocation.

Code would translate to:

```scala
class ControllerInstanceId(val asString: String) extends AnyVal
```

I'm wondering if we should fail out early if the controller name containers bad characters vs. implicitly removing them. Any bad characters that kube/mesos add by default?

Reply:
Don't think anything kube adds any bad characters.  The argument passed in on the command line as the controller name will be a DNS-legal hostname [a-z0-9.].  Invoker ids will either by the ipaddress of the workernode [0-9.] or a DNS legal hostname like invoker0 [a-z0-9.].

Reply:
My take was to leave the sanitization out completely. Was wondering if we should just leave it fail on topic creation and tell the operator to choose a better name then?

Reply:
How about this resolution: https://github.com/apache/incubator-openwhisk/commit/3b9e52f1d82ef7159ba51b3d83fcc161afee129b

Simplifies the name checking as @markusthoemmes suggested (no sanitization) and still checks the controller instance consists of valid characters.